packages/ui-components/src/icon/jlicon.tsx

```diff
@@ -0,0 +1,238 @@
+// Copyright (c) Jupyter Development Team.
+// Distributed under the terms of the Modified BSD License.
+
+import React from 'react';
+import ReactDOM from 'react-dom';
+
+import { iconStyle, IIconStyle } from '../style/icon';
+import { getReactAttrs, classes, classesDedupe } from '../utils';
+
+export class JLIcon {
+  private static _instances = new Map<string, JLIcon>();
+
+  static get(name: string): JLIcon {
+    return JLIcon._instances.get(name) as JLIcon;
+  }
+
+  constructor({ name, svgstr, _debug = false }: JLIcon.IOptions) {
+    this.name = name;
+    this._svgstr = svgstr;
+    this._debug = _debug;
+
+    this.react = this._initReact();
+
+    JLIcon._instances.set(name, this);
+  }
+
+  class({ className, ...propsStyle }: { className?: string } & IIconStyle) {
+    return classesDedupe(className, iconStyle(propsStyle));
+  }
+
+  element({
+    className,
+    container,
+    title,
+    tag = 'div',
+    ...propsStyle
+  }: JLIcon.IProps = {}): HTMLElement | null {
+    // ensure that svg html is valid
+    const svgElement = this.resolveSvg();
+    if (!svgElement) {
+      // bail if failing silently
+      return null;
+    }
+
+    // create a container if needed
+    container = container || document.createElement(tag);
+
+    this._initContainer({ container, className, propsStyle, title });
+
+    // add the svg node to the container
+    container.appendChild(svgElement);
+
+    return svgElement;
+  }
+
+  replaceSvg(svgstr: string) {
+    this._svgstr = svgstr;
+  }
+
+  render(host: HTMLElement, props: JLIcon.IProps = {}): void {
+    return ReactDOM.render(<this.react container={host} {...props} />, host);
+  }
+
+  resolveSvg(title?: string): HTMLElement | null {
+    const svgDoc = new DOMParser().parseFromString(
+      this._svgstr,
+      'image/svg+xml'
+    );
+    const svgElement = svgDoc.documentElement;
+
+    if (svgElement.getElementsByTagName('parsererror').length > 0) {
+      const errmsg = `SVG HTML was malformed for icon name: ${name}`;
+      // parse failed, svgElement will be an error box
+      if (this._debug) {
+        // fail noisily, render the error box
+        console.error(errmsg);
+        return svgElement;
+      } else {
+        // bad svg is always a real error, fail silently but warn
+        console.warn(errmsg);
+        return null;
+      }
+    } else {
+      // parse succeeded
+      if (title) {
+        Private.setTitleSvg(svgElement, title);
+      }
+
+      return svgElement;
+    }
+  }
+
+  string() {
+    return this._svgstr;
+  }
+
+  unrender(host: HTMLElement): void {
```

Comment:
Maybe make static?

Reply:
Can't. The `render/unrender` methods exist so you can pass instances of `JLIcon` to Lumino (via the `title.iconRenderer` param). The methods are there so that Lumino can treat any `JLIcon` instance as an `IRenderer`, so they need to be public instance methods.

My current thinking is that client code in general should not directly call `render` or `unrender`.

Reply:
Ah I see, that makes sense!